### PR TITLE
Add GitLab client provider to support dynamic credentials fetching

### PR DESCRIPTION
**Pull request by Celdir**
https://github.com/abcxyz/team-link/issues/70

### CODE REVIEW

File: pkg/gitlab/client.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+// ClientProvider provides a GitLab client.
+type ClientProvider struct {
+	httpClient  *http.Client
+	instanceURL string
+	keyProvider github.KeyProvider
+}
+
+// NewGitLabClientProvider creates a new GitLabClientProvider.
+func NewGitLabClientProvider(httpClient *http.Client, instanceURL string, keyProvider github.KeyProvider) *ClientProvider {
```
Comment (lock14) [inline pkg/gitlab/client.go:35]: I don't like that we are using the `github` package inside the `gitlab` one. Is it possible to refactor the `KeyProvider` into a common package they can both use?

**Review by lock14 — CHANGES_REQUESTED** [review]

File: pkg/gitlab/client.go (hunk 1 above)
Comment (sailorlqh) [reply pkg/gitlab/client.go:35]: What do you think if we move all the client/provider constructor into `pkg/client/provider.go`, in there we can have `NewGitLabClient`, `NewGoogleGroupClient`, `NewGitHubClient`

Comment (lock14) [reply pkg/gitlab/client.go:35]: I may be misunderstanding your suggestion, but I think Client constructors should stay in their respective packages as it is likely we will want to access package private fields/vars when constructing clients.

My suggestion was specifically about the `KeyProvider` which is a pretty simple interface that just returns a sequence of bytes and does not specifically have anything to do with `GitHub` or `GitLab`.

Comment (sailorlqh) [reply pkg/gitlab/client.go:35]: Sg, thanks!

Comment (Celdir) [reply pkg/gitlab/client.go]: Moved `KeyProvider` to a new package called `credentials`.

**Review by lock14 — APPROVED** [review]

**Review by sailorlqh — APPROVED** [review]

**Pull request merged**